File: androidApp/src/main/java/com/ruizurraca/carapp/SyncStatusIndicatorView.kt
package com.ruizurraca.carapp

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import com.ruizurraca.carapp.core.common.SyncStatus
import com.ruizurraca.carapp.core.common.UiMessage

/**
 * Renders the shared backup status without deriving synchronization state in the host.
 * The visible label owns the localized accessibility description, the colored dot is decorative,
 * Failed is the only state with Retry, and a typed retry error is rendered through [ErrorText] only
 * beside Failed, because `§9.9` reserves the error presentation for it.
 */
@Composable
internal fun SyncStatusIndicator(
    status: SyncStatus,
    message: UiMessage?,
    onRetry: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val visual = syncStatusVisual(status)
    val isError = syncStatusIsError(visual)
    val label = stringResource(syncStatusLabelResource(visual))
    val description = stringResource(R.string.backup_status_description, label)
    Column(
        modifier =
            modifier
                .fillMaxWidth()
                .padding(horizontal = 20.dp, vertical = 8.dp),
        verticalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        Row(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .testTag(SyncStatusTestTags.INDICATOR),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Box(
                modifier =
                    Modifier
                        .size(8.dp)
                        .background(
                            color = if (isError) MaterialTheme.colorScheme.error else MaterialTheme.colorScheme.primary,
                            shape = CircleShape,
                        ),
            )
            Text(
                text = label,
                modifier =
                    Modifier
                        .weight(1f)
                        .semantics { contentDescription = description },
                style = MaterialTheme.typography.labelMedium,
                color = if (isError) MaterialTheme.colorScheme.error else MaterialTheme.colorScheme.onSurfaceVariant,
            )
            if (isError) {
                TextButton(
                    onClick = onRetry,
                    modifier = Modifier.testTag(SyncStatusTestTags.RETRY),
                ) {
                    Text(stringResource(R.string.backup_status_retry))
                }
            }
        }
        // The holder withdraws the message when the status leaves Failed (`§14`); this guard also
        // covers the turn in which the vehicle-list relay has moved on and the sync holder has not.
        if (isError) {
            ErrorText(message, testTag = SyncStatusTestTags.ERROR)
        }
    }
}

object SyncStatusTestTags {
    const val INDICATOR = "backup_status_indicator"
    const val RETRY = "backup_status_retry"
    const val ERROR = "backup_status_error"
}

File: iosApp/SyncStatusIndicatorView.swift
import SwiftUI
import Shared

/// The discreet backup status indicator of `docs/SPECIFICATION.md §3.1`, drawn as the status chip the
/// iOS design already places on this screen (`docs/DESIGN.md §4`, screen 02 home).
///
/// It renders the value the shared layer publishes and nothing else: the precedence and the
/// connectivity rule belong to `docs/CONTRACTS.md §9.9`, and `§14` forbids this view from computing a
/// second status. The retry affordance appears only for the visual `§9.9` defines as an error, and it
/// calls the existing `SyncStateHolder.retryFailed()`, so a failure it produces arrives as the
/// holder's typed `UiMessage` and is rendered through the one existing mapping, `localizedText`,
/// and only beside the `Failed` visual, because `§9.9` reserves the error presentation for it.
///
/// The status text owns the accessible state; the coloured dot is decorative. The Retry button is
/// deliberately left as its own element: combining the row into one accessibility element would fold
/// a button into a non-actionable node and make manual recovery unreachable with a screen reader.
struct SyncStatusIndicatorView: View {
    let status: SyncSyncStatus
    let message: UiMessage?
    let onRetry: () -> Void

    private var visual: SyncStatusVisual { syncStatusVisual(status) }

    private var label: String {
        String(localized: String.LocalizationValue(syncStatusLabelKey(visual)))
    }

    private var isError: Bool { syncStatusIsError(visual) }

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            HStack(spacing: 8) {
                // Decorative: the adjacent text carries the accessible status.
                Circle()
                    .fill(isError ? Color.red : Color.accentColor)
                    .frame(width: 8, height: 8)
                    .accessibilityHidden(true)

                Text(label)
                    .font(.footnote)
                    .foregroundColor(isError ? .red : .secondary)
                    .accessibilityIdentifier("backup_status_indicator")
                Spacer()
                if isError {
                    Button(String(localized: "backup_status_retry"), action: onRetry)
                        .font(.footnote)
                        .accessibilityIdentifier("backup_status_retry")
                }
            }

            if isError, let message {
                Text(message.localizedText)
                    .font(.footnote)
                    .foregroundColor(.red)
                    .accessibilityIdentifier("backup_status_error")
            }
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 8)
    }
}

/// Draws the indicator from the process model and observes it.
///
/// `VehicleListView` holds `WalkingSkeletonModel` as a plain reference. SwiftUI does not re-evaluate
/// a child whose stored references are unchanged, so a status change or a retry failure that
/// arrived without a vehicle-list change was never drawn. This row is the only observer of the model
/// on that screen, so a sync-only change redraws the indicator and nothing else. The status source
/// stays `SyncStateHolder.state`, as `docs/adr/0193` assigns to iOS.
struct BackupStatusRow: View {
    @ObservedObject var model: WalkingSkeletonModel

    var body: some View {
        SyncStatusIndicatorView(
            status: model.syncState.status,
            message: model.syncState.message,
            onRetry: model.retryBackup
        )
    }
}
